lucet-runtime/lucet-runtime-internals/src/future.rs

```diff
@@ -0,0 +1,189 @@
+
+/// This is the same type defined by the `futures` library, but we don't need the rest of the
+/// library for this purpose.
+type LocalBoxFuture<'a, T> = Pin<Box<dyn Future<Output = T> + 'a>>;
+
+/// A unique type that wraps a boxed future with a boxed return value.
+///
+/// Type and lifetime guarantees are maintained by `Vmctx::block_on` and `Instance::run_async`. The
+/// user never sees this type.
+struct YieldedFuture(LocalBoxFuture<'static, ResumeVal>);
+
+/// A unique type for a boxed return value. The user never sees this type.
+struct ResumeVal(Box<dyn Any + Send + 'static>);
+
+impl Vmctx {
+    /// Block on the result of an `async` computation from an instance run by `Instance::run_async`.
+    ///
+    /// Lucet hostcalls are synchronous `extern "C" fn` functions called from WebAssembly. In that
+    /// context, we cannot use `.await` directly because the hostcall is not `async`. While we could
+    /// block on an executor using `futures::executor::block_on` or
+    /// `tokio::runtime::Runtime::block_on`, that has two drawbacks:
+    ///
+    /// - If the Lucet instance was originally invoked from an async context, trying to block on the
+    ///   same runtime will fail if the executor cannot be nested (all executors we know of have this
+    ///   restriction).
+    ///
+    /// - The current OS thread would be blocked on the result of the computation, rather than being
+    ///   able to run other async tasks while awaiting. This means an application will need more
+    ///   threads than otherwise would be necessary.
+    ///
+    /// This `block_on` operator instead yields yields a future back to a loop that runs in
+    /// `Instance::run_async`, which `.await`s on it and then resumes the instance with the
+    /// result. The future runs on the same runtime that invoked `run_async`, avoiding problems of
+    /// nesting, and allowing the current OS thread to continue performing other async work.
+    ///
+    /// Note that this method may only be used if `Instance::run_async` was used to run the VM,
+    /// otherwise it will terminate the instance with `TerminationDetails::AwaitNeedsAsync`.
+    pub fn block_on<'a, R>(&'a self, f: impl Future<Output = R> + 'a) -> R
+    where
+        R: Any + Send + 'static,
+    {
+        // Die if we aren't in Instance::run_async
+        match self.instance().state {
+            State::Running { async_context } => {
+                if !async_context {
+                    panic!(TerminationDetails::AwaitNeedsAsync)
+                }
+            }
+            _ => unreachable!("Access to vmctx implies instance is Running"),
+        }
+        // Wrap the Output of `f` as a boxed ResumeVal. Then, box the entire
+        // async computation.
+        let f = Box::pin(async move { ResumeVal(Box::new(f.await)) });
+        // Change the lifetime of the async computation from `'a` to `'static.
+        // We need to lie about this lifetime so that `YieldedFuture` may impl
```

Comment:
Mind adding a note about why this is safe? AIUI the explanation is basically: the relevant stack frame (that `'a` is tied to) is essentially frozen in place during this process.